_gitbook/syntax_and_semantics/generics.md

```diff
@@ -1 +1,78 @@
+```
+Array(String).new # or "[] of String"
+Array(Symbol).new # or "[] of Symbol"
+Array(Array).new # or "[] of Array"
+```
+
+*Note*: An array (or hash, or range) literal with elements in it is a bit of a special case. It will automatically have its type set to whatever it contains without having to declare it explicitly.
+
+The type of T is important here for at least two reasons:
+In Array's case, the class is implemented by a Pointer class (actually a struct, but that's another story) that is itself generic, so the type of the Array will also be the type of its internal Pointer.
+More relevant to the subject of generic classes, many methods accept only objects that are of the array's type and a method missing exception will be raised (at compile time) if given anything else.
+Example:
+```
+class Array(T)
+  def unshift(obj : T)
+    # body omitted
+  end
+end
+```
+Everywhere in the definition of a generic class you can use that "T" to reference the type of the class. In this case, it's used to define a method which will only accept an argument of type T, whatever that might be.
+Let's see what happens if you try to call #unshift with an object that is not of the array's type:
+```
+  array = Array(String).new
+  array.unshift(5)
+
+  # no overload matches 'Array(String)#unshift' with types Int32
+  # Overloads are:
+  # - Array(String)#unshift(obj : String)
+```
+A generic class can also have an union of two or more types as a type variable, which is written in this way:
+`Array(String | Symbol).new # or "[] of String | Symbol"`
+When calling a method on an any element of an array such as the one above the compiler will raise a method missing exception unless both String and Symbol implement it:
+`[:abc, "bcd"].first.reverse # undefined method 'reverse' for Symbol`
+This is because as far as the compiler is concerned, the return value of `#first` could be either a String or a Symbol and Symbol doesn't implement a #reverse method.
+A method can also explicitly reference an "unknown" type that is not necessarily the class' type. A good example is #map:
+```
+def map(&block : T -> U)
+  ary = [] of U
+  each { |e| ary << yield e }
+  ary
+end
+```
+That block argument can be read as "a block that yields elements of type T and returns elements of type U". As with "T", the name "U" is a convention, totally arbitrary. #map then is able to create an array of the type that the block returns which might be different from T.
+A generic class can have more than one type, like Hash:
+```
+class Hash(K, V)
+end
+```
+In this case K stands for the type of the keys, and V for the type of the values.
+```
+hash = Hash(Symbol, String) # or "{} of Symbol => String"
+hash[:name] = "Exilor"
+```
+Range too used to have two and will have two again when 0.6.2 is released:
```

Comment:
I'll try to reuse what you have written and what I know about generics to write this page. There are many things to explain, for example type argument inference from new/initialize, and why there's a need for generics.